### ISSUE TITLE
Integrals slower with assumptions on variables

### ISSUE THREAD

**Issue by oscargus** [sympy/sympy#17114]
After #17093 
```
test_issue_14782 - Took 30.243 seconds
test_issue_7161 - Took 37.741 seconds
test_issue_11876 - Took 67.013 seconds
test_W4 - Took 32.775 seconds
test_W5 - Took 37.324 seconds
test_W3 - Took 37.335 seconds
```

All these are pretty simple integrals with numerical bounds, which did not show up among the slowest tests before this.

Looking at 
```
def test_issue_14782():
    f = sqrt(-x**2 + 1)*(-x**2 + x)
    assert integrate(f, [x, -1, 1]) == - pi / 8
    assert integrate(f, [x, 0, 1]) == S(1) / 3 - pi / 16
```
it is the second integral that takes long time. In fact, all integrals, except W4, has one bound that is 0. Maybe that is a clue.

W4 is rewritten to 
![image](https://user-images.githubusercontent.com/8114497/60259334-51d7d280-98d7-11e9-99c0-967ab97eb699.png)
which avoid any further integration. Removing the abs leads to a very fast (and correct) result.

Any ideas @smichr ?

**Comment by oscargus** [comment]
Turns out that the W4 integral can be evaluated with another `doit()`. Still doesn't really explain the difference in time.

**Comment by smichr** [comment]
> Any ideas @smichr ?

try profiling to see where it is getting stuck? When I broke the running code for `integrate(sqrt(1 - x**2)*(-x**2 + x),(x,0,1))` it was in `as_ordered_terms` -- not sure why that step is necessary -- and another time in `solveset`.

**Comment by smichr** [comment]
I notice, too, when printing out the integrand in `add_rule` this difference:
```python
>>> integrate(sqrt(1 - x**2)*(-x**2 + x),(x,0,1))
-x**2*sqrt(1 - x**2) + x*sqrt(1 - x**2)
1/8 - cos(4*_theta)/8
-x**2*sqrt(1 - x**2) + x*sqrt(1 - x**2)
1/8 - cos(4*_theta)/8
1/3 - pi/16

>>> integrate(sqrt(1 - x**2)*(-x**2 + x),(x,0,1))
2*sqrt(1 - 1/_u**2)/_u**3 - 2*sqrt(1 - 1/_u**2)/_u**4
2*_u**4/sqrt(1 - _u**2) + 2*_u**2 - 2*_u**2/sqrt(1 - _u**2)
1/2 - cos(2*_theta)/2
cos(2*_theta)**2/4 - cos(2*_theta)/2 + 1/4
cos(4*_theta)/2 + 1/2
cos(2*_theta)**2/4 - cos(2*_theta)/2 + 1/4
cos(4*_theta)/2 + 1/2
...
much repetition follows
```

So there is some u-substitution that is triggered by the assumptions that probably should not be happening.

Changing assumptions to account for 0 did not help:
```python
-            if all(i.is_nonnegative for i in l) and not x.is_nonnegative:
-                d = Dummy(positive=True)
-            elif all(i.is_nonpositive for i in l) and not x.is_nonpositive:
+            if all(i.is_negative for i in l) and not x.is_negative:
+                d = Dummy(negative=True)
+            elif all(i.is_positive for i in l) and not x.is_positive:
                 d = Dummy(negative=True)
+            elif all(i.is_nonnegative for i in l) and not x.is_nonnegative:
+                d = Dummy(nonnegative=True)
+            elif all(i.is_nonpositive for i in l) and not x.is_nonpositive:
+                d = Dummy(nonpositive=True)
```

**Comment by oscargus** [comment]
Yes, profiling is of course a good idea. Seems counterintuitive since one would guess that assumptions would just make it easier somehow. But then, it obviously depends on how the code is written.

**Comment by oscarbenjamin** [comment]
The integrals code is very sensitive to changes in the input that lead to different choices of algorithm etc so I wouldn't guess that something like the change here would make things easier in *all* cases.

**Comment by oscargus** [comment]
No clearly it doesn't. However, it seems like it should, fundamentally, be easier the more one knows.

**Comment by oscargus** [comment]
I've looked a bit more into the integral in W4.

Turns out that as the integral variable is nonnegative there are many more possible ways to do manual integration. Hence, it takes much longer time to generate these.

**Comment by smichr** [comment]
Is that what is happening with `integrate(f, [x, 0, 1])`? My comment [here](https://github.com/sympy/sympy/issues/17114#issuecomment-506489252) makes me think that some sort of caching for doit should be used so the same expressions aren't tested over and over. A casual observation of the results that I refer to suggests that there is a lot of repetition.

**Comment by oscargus** [comment]
Yes. I spent some time trying to make it print properly/readable, but didn't manage.

To recreate run `integral_steps` from manualintegrate which returns all possible alternatives.

I guess that it may also make sense to simply try one combination first and if that gives a "trivial" result just skip the rest. Now there are lot of different results even for a relatively simple integral.

I also noted that there were `Piecewise`s with a single piece involved (x < 1) that possibly may be evaluated earlier. Not sure how though. Lots of call to piecewise_fold, although most are for expressions that cannot be folded.

Also, there are substitutions later where assumptions are not propagated. Now, it looks like it is not always a good idea to propagate it, but I guess that in the end it should still be that.
